### ISSUE TITLE
Grapher redesign

### ISSUE THREAD

**Issue by adamerose** [adamerose/PandasGUI#121]
I'm making a lot of changes to the Grapher right now and this issue will track it. The main change is pulling all the functions that create Plotly figures out of `grapher.py` and putting them into a separate module called `jotly.py`. This will be the single module where all plots are defined along with all their options. I've redesigned the `Grapher` and `Dragger` (which I renamed to `FuncUi`) to import functions from `jotly.py` and automatically generate a UI from the function's arguments using the typehints, this will be much easier to maintain since all changes to the plotting function arguments will be reflected in the autogenerated UI schema. 

Example:

![image](https://user-images.githubusercontent.com/10884874/111572337-c7a46180-877e-11eb-9c80-ed3e6a53f235.png)

The old drag n drop functionality is still there, but is redesigned and the code is simpler now. Instead of a QTreeWidget with root level QTreeWidgetItems for argument names and child QTreeWidgetItems under each one for argument values, it is flattened to only one level of QTreeWidgetItems where the first column is argument name and the 2nd column is a widget to set and display the value.

Here's how the `jotly` type hints map to PyQt widgets:

- `ColumnName` (which is just a `str`) -> A textbox that you can type in or drop columns onto
- `Literal` -> A combobox dropdown with options for the valid values defined in the type hint
- `bool` -> A checkbox
- (not implemented yet) `ColumnNameList` (which is `List(str)`) -> A list of multiple text boxes where you can drop column names and add/remove rows.

The reason I defined `ColumnName` as a subclassed type is to differentiate arguments that should be a DataFrame column name from ones that can be any string, like `title`.

Other new additions
- Added a console to display errors generating plots
- Added a button `Preview Kwargs` to show your args as defined by the UI in a text box as a dict
- Automatically re-render the plot any time the arguments are changed

And some more things I still need to add
- Re-add the old logic like sorting and aggregating for line/bar plots and more args to jotly
- Implement `ColumnNameList` for args that accept a list of column names
- Heuristics to differentiate categorical and continuous numeric columns. Should add a `color_mode` arg like [this](https://plotly.com/python/discrete-color/#discrete-colors-in-dash), but add a third `auto` option that uses a heuristic.
- Sync selections made between the Grapher and DataFrameExplorer. JMP does this and it's very cool: https://www.screencast.com/t/yKmLTaFaP9
- In addition to facet_row and facet_col, I might want these options:
  - facet_tab: interactive tabs you can click between
  - facet_wrap: splits into subplots like facet_row and facet_col but using a single variable and wraps them all into a square grid. This should be easy because plotly already provides facet_col_wrap, I will just make that number auto-calculated to give a grid.
  - facet_page: like facet_col but each subplot is full size and you can scroll through them
  These were partly inspired by JMP. ([page ](https://www.screencast.com/t/PuQHKwZG)/ [wrap](https://www.screencast.com/t/YaITlhzUe))
 - Allow opening Grapher to a specific state. So maybe you could type `x = pandasgui.show_grapher(df, type='scatter', args={ 'x': 'age', 'y': 'fare', 'color':'survived''})` and it will pop open like [this](https://www.screencast.com/t/2qnzQdhnkgS)
 - Make Grapher a dialog instead of a Tab so you can have multiple Graphers open at once for a single DataFrame
- Add checkbox to disable automatic re-rendering
- Fix the GroupBox title styling

I temporarily broke some things:

- Automatic title generation. I will move this to a utility function within jotly functions, it will be easier to maintain and understand things if there's no magic happening in the Grapher adding extra kwargs to pass to the plot functions that the user didn't define in the UI.
- Render mode. Also should be moved to jotly. But I also don't understand the point of this, Plotly docs say this:

> 'webgl' is likely necessary for acceptable performance above 1000 points but rasterizes part of the output. auto' uses heuristics to choose the mode.

@fdion Were the `auto` heuristics failing? Why did you add an override for certain plots?

**Comment by fdion** [comment]
@adamerose need to know when you think things will settle down on the major refactoring so I can address the automated title/render mode if it is still a regression. I'd like to get to the point where a new pypi release can be made so it'll have the grapher splitter change .

And more generally, when it makes sense to address some of the other things, as I was waiting for the dust to settle a bit! :)

**Comment by adamerose** [comment]
> @adamerose need to know when you think things will settle down on the major refactoring so I can address the automated title/render mode if it is still a regression. I'd like to get to the point where a new pypi release can be made so it'll have the grapher splitter change .
> 
> And more generally, when it makes sense to address some of the other things, as I was waiting for the dust to settle a bit! :)

Probably soon. There are just a couple things that I've been scratching my head over....

- The best way to allow niche settings. For example I think `render_mode` should not appear in the Grapher UI by default, almost all users will want WebGL on and I don't want a cluttered UI with things only a few people will need, but I also don't want to alienate users that request features like this. 
  - A possible solution is have args in Jotly functions with leading underscores for things I don't want to be included by default in the autogenerated UI, so in FuncUi I will just skip the UI generation on kwargs with leading underscores but the user can still use them with the custom kwargs dialog
  - Another solution would be to use a special type hint for them like HiddenOption that FuncUi ignores. This would mean no useful type hint. I'm leaning towards the first option unless you have an aversion to args having leading underscores and have a suggested alternative
- The SETTINGS_STORE is getting cluttered with settings for default Grapher values, so I plan to move these all inside a single dict
- I'll be re-adding your title generating function soon

**Comment by fdion** [comment]
One way to handle **all** plotly settings would be to allow a kwarg dict to show/pandasgui: pass them all the way to jotly if they come from the initial call. Internally, you can handle that with either method you propose. Once in jotly, globally for all plots, any kwargs that start with layout_ are applied to update_layout, anything else to update_trace. And like you said, totally skip the UI.

This would be extremely flexible. New option in plotly? no problem, already supported.

**Comment by adamerose** [comment]
Pushed my changes so far. Things are mostly working and I put back in the title generation but it needs to be fixed up since I renamed `apply_mean` to `aggregation` and removed `apply_sort` (I think it's [not needed](https://github.com/adamerose/PandasGUI/issues/103#issuecomment-808674956)). Render mode is no longer an option inside the Grapher it's just an option in Preferences and automatically applies to the Grapher where needed

**Comment by fdion** [comment]
Just had to remove `apply_sort` from settings and rename `apply_mean` to aggregation and pass 'none' instead of False, and it started up.

Some feedback:

- drag and drop to plot setup boxes seems easier to do than to drop on trees in the previous dragger setup
- haven't dropped a variable outside the designated spots (would happen randomly with the tree setup)
- nice touch adding the `marginal`, `cumulative` and `trendline` as UI options
- drag and drop multiple variables on x or y etc is definitely needed, and it is on the TODO (ColumnNameList)
- i see you don't have to click finish before it generates a plot, for small data sets not a problem, but for large (> 200,000 data points) it slows down the process a  good bit. I see you have that also in the TODO (auto render checkbox)
- x and y are inverted in the dragger
- stuff like `text`, `markers`, `hover_name` etc missing.
- still would need `apply_sort` for some sequential experiments data. Another approach: if `apply_sort` was instead `sort` and would take a column name(s) and sort by that?
- I haven't had the chance to fully test the automated `title`

**Comment by fdion** [comment]
In case you are wondering, number 1 hurdle at this time is:  
- Implement ColumnNameList for args that accept a list of column names

**Comment by adamerose** [comment]
Yeah still haven't had time for that, will probably do it by this weekend

**Comment by adamerose** [comment]
Done. Only enabled it on Splom (scatter_matrix) so far. 

![image](https://user-images.githubusercontent.com/10884874/115345743-365d4c00-a17d-11eb-9281-285f3b12fea6.png)


It'll probably months to finish all the ideas in this thread - can you give a shortlist of what else you think needs to be done before putting out another release makes sense?

**Comment by adamerose** [comment]
Done. Only enabled it on Splom (scatter_matrix) so far. 

![image](https://user-images.githubusercontent.com/10884874/115345743-365d4c00-a17d-11eb-9281-285f3b12fea6.png)

It'll probably take months to finish all the ideas in this thread - can you give a shortlist of what else you think needs to be done before putting out another release makes sense? I'm never in a hurry to put out new releases but I know you mentioned it above and can prioritize some things

**Comment by fdion** [comment]
Hey that's great. Let me look at this, I'll let you know.

**Comment by fdion** [comment]
I ended up packaging a wheel file with the pre-redesign code but with graph splitter, to buy some time, so not super urgent doing a release ATM.

I've been looking into the `ColumnNameList`, and not having much success getting it to work for X and Y on line. Looked at Wordcloud and SPLOM and it should `just work`, but get a `NoneType` error on `set_names` (`func_ui`).

**Comment by fdion** [comment]
Another PR for this ticket: https://github.com/adamerose/PandasGUI/pull/137
